test/box-luatest/gh_9247_bsize_on_concurrent_dml_and_alter_test.lua

```diff
@@ -0,0 +1,95 @@
+local server = require('luatest.server')
+local t = require('luatest')
+local g = t.group('gh-9247', {{rollback = false}, -- test box.commit()
+                              {rollback = true}}) -- test box.rollback()
+
+g.before_all(function(cg)
+    cg.server = server:new()
+    cg.server:start()
+end)
+
+g.after_all(function(cg)
+    cg.server:drop()
+end)
+
+g.after_each(function(cg)
+    cg.server:exec(function()
+        box.space.test:drop()
+    end)
+end)
+
+-- Check that space:bsize() is correct for a built-in-background primary index
+-- with concurrent inserts.
+-- Based on `test/box/alter-primary-index-tuple-leak-long.test.lua'.
+g.test_concurrent_replace_and_alter = function(cg)
+    cg.server:exec(function(rollback)
+        local fiber = require('fiber')
+        local s = box.schema.space.create('test')
+        local idx = s:create_index('pk')
+        local alter_started = false
+        local bsize_after_replace = 0
+        local data = string.rep('a', 66)
+
+        -- Create a table with enough tuples to make primary index alter in
+        -- background.
+        for i = 0, 7000 do
+            s:insert{i, i, data}
+        end
+
+        -- Wait for fiber yield during altering and insert some tuples.
+        local fib_replace = fiber.new(function()
+            while not alter_started do
+                fiber.sleep(0)
+            end
+            box.begin()
+            for i = 8000, 9000 do
+                s:insert{i, i, data}
+            end
+            box.commit()
+            bsize_after_replace = s:bsize()
+        end)
+
+        -- Alter primary index.
+        -- If the index will not be altered in background, the test will fail
+        -- after timeout (fib_replace:join() will wait forever).
+        local fib_alter = fiber.new(function()
+            alter_started = true
+            box.begin()
+            idx:alter{parts = {{field = 2}}}
+            if rollback then
+                box.rollback()
+            else
+                box.commit()
+            end
+            alter_started = false
+        end)
+
+        fib_replace:set_joinable(true)
+        fib_alter:set_joinable(true)
+        fib_replace:join()
+        fib_alter:join()
+
+        local bsize_after_alter = s:bsize()
+        t.assert_equals(bsize_after_replace, bsize_after_alter)
```

Comment:
The check isn't good enough. You should check that the sum `tuple.bsize` equals `space.bsize`.

Reply:
Done.